go/vt/vtgate/planbuilder/operator_transformers.go

```diff
@@ -771,7 +771,9 @@ func updateSelectedVindexPredicate(op *operators.Route) sqlparser.Expr {
 		if !ok {
 			continue
 		}
-
+		if sqlparser.Equals.Expr(cmp.Right, sqlparser.ListArg(engine.DmlVals)) {
+			continue
+		}
```

Comment:
I think there can be a better solution for this.

Reply:
This is a very fragile hack.

Reply:
We can ignore if right is an argument type

Reply:
> We can ignore if right is an argument type

I tried making this modification, 
```
func updateSelectedVindexPredicate(routing operators.Routing) sqlparser.Expr {
	tr, ok := routing.(*operators.ShardedRouting)
	if !ok || tr.Selected == nil {
		return nil
	}

	_, isMultiColumn := tr.Selected.FoundVindex.(vindexes.MultiColumn)
	for idx, expr := range tr.Selected.Predicates {
		cmp, ok := expr.(*sqlparser.ComparisonExpr)
		if !ok || cmp.Operator != sqlparser.InOp {
			continue
		}
		_, ok = cmp.Left.(*sqlparser.ColName)
		if !ok {
			continue
		}
		_, ok = cmp.Right.(sqlparser.ListArg)
		if ok {
			continue
		}
		var argName string
		if isMultiColumn {
			argName = engine.ListVarName + strconv.Itoa(idx)
		} else {
			argName = engine.ListVarName
		}

		cmp.Right = sqlparser.ListArg(argName)
	}
	return nil
}
```

but it caused some other cases in plan_test to fail, just like this one.
```
        Diff:
        {
            "Instructions": {
                "Inputs": [
                    {
                        "Aggregates": "any_value(0) AS 1, sum_count(1) AS count(id)",
                        "Inputs": [
                            {
                                "Inputs": [
                                    {
                                        "FieldQuery": "select `name`, keyspace_id from name_user_vdx where 1 != 1",
                                        "Keyspace": {
                                            "Name": "user",
                                            "Sharded": true
                                        },
                                        "OperatorType": "Route",
                                        "Query": "select `name`, keyspace_id from name_user_vdx where `name` in ::__vals" => "select `name`, keyspace_id from name_user_vdx where `name` in ::name",
                                        "Table": "name_user_vdx",
                                        "Values": [
                                            "::name"
                                        ],
                                        "Variant": "IN",
                                        "Vindex": "user_index"
                                    },
                                    {
                                        "FieldQuery": "select 1, count(id) from `user` where 1 != 1",
                                        "Keyspace": {
                                            "Name": "user",
                                            "Sharded": true
                                        },
                                        "OperatorType": "Route",
                                        "Query": "select 1, count(id) from `user` where `name` = 'a'",
                                        "Table": "`user`",
                                        "Variant": "ByDestination"
                                    }
                                ],
                                "Keyspace": {
                                    "Name": "user",
                                    "Sharded": true
                                },
                                "OperatorType": "VindexLookup",
                                "Values": [
                                    "'a'"
                                ],
                                "Variant": "Equal",
                                "Vindex": "name_user_map"
                            }
                        ],
                        "OperatorType": "Aggregate",
                        "Variant": "Scalar"
                    }
                ],
                "OperatorType": "Filter",
                "Predicate": "count(id) = 10",
                "ResultColumns": 1
            },
            "Original": "select 1 from user having count(id) = 10 and name = 'a'",
            "QueryType": "SELECT",
            "TablesUsed": [
                "user.user"
            ]
        }
```

Reply:
The problem is in the `routing` engine primitive which assumes the bind variables always to have the prefix `__vals`. Only if we change that assumption can we fix this.
For now, we can keep this change